src/index.ts

```diff
@@ -147,31 +186,74 @@ export class CSSStyleObserver {
+   * @param processChange Function to process each changed property
+   * @returns Collected changes
+   */
+  private _processObservedVariables<T>(
+    computedStyle: CSSStyleDeclaration,
+    processChange: (propertyName: string, currentValue: string, previousValue: string, hasChanged: boolean) => T
+  ): T[] {
+    const changes: T[] = [];
+
+    this._observedVariables.forEach((propertyName) => {
+      const currentValue = computedStyle.getPropertyValue(propertyName);
+      const previousValue = this._cachedValues[propertyName] || '';
+      const hasChanged = currentValue !== previousValue;
+
+      if (this._notificationMode === NotificationMode.ALL || hasChanged) {
+        changes.push(processChange(propertyName, currentValue, previousValue, hasChanged));
+        this._cachedValues[propertyName] = currentValue;
+      }
+    });
+
+    return changes;
+  }
+
+  /**
+   * Handlers for return formats
+   */
+  private _returnFormatHandlers: Record<ReturnFormat, (computedStyle: CSSStyleDeclaration) => CSSDeclarations | { [key: string]: CSSPropertyInfo }> = {
+    [ReturnFormat.VALUE_ONLY]: (computedStyle) => {
+      const changes = this._processObservedVariables(computedStyle, (propertyName, currentValue) => ({
+        [propertyName]: currentValue
+      }));
+
+      return Object.assign({}, ...changes);
+    },
+    [ReturnFormat.OBJECT]: (computedStyle) => {
+      const changes = this._processObservedVariables(computedStyle, (propertyName, currentValue, previousValue, hasChanged) => ({
+        [propertyName]: {
+          value: currentValue,
+          previousValue,
+          changed: hasChanged,
+        },
+      }));
+
+      return Object.assign({}, ...changes);
+    },
+  };
+
   /**
    * Collect CSS variable values and invoke callback.
    */
   private _handleUpdate(): void {
     if (this._targetElement) {
       const computedStyle = getComputedStyle(this._targetElement);
 
-      const changedProperties: CSSDeclarations = {};
-
-      this._observedVariables.forEach(propertyName => {
-        const currentValue = computedStyle.getPropertyValue(propertyName);
-        const previousValue = this._cachedValues[propertyName] || '';
-
-        if (this._notificationMode === NotificationMode.ALL || currentValue !== previousValue) {
-          changedProperties[propertyName] = currentValue;
-          this._cachedValues[propertyName] = currentValue;
-        }
-      });
-
-      // Invoke callback only if there are changes
-      if (Object.keys(changedProperties).length > 0) {
-        this._callback(changedProperties);
+      // Execute the handler for the current return format, default to ReturnFormat.VALUE_ONLY if not valid
+      const handler = this._returnFormatHandlers[this._returnFormat] ?? this._returnFormatHandlers[ReturnFormat.VALUE_ONLY];
+      const changes = handler(computedStyle);
+      
+      if (Object.keys(changes).length > 0) {
+        this._callback(changes);
```

Comment:
Ideally, this would become this:

```suggestion
      const changes = this. _processObservedVariables(computedStyle);
      if (Object.keys(changes).length == 0) return;
      
      const format = this._returnFormatHandlers[this._returnFormat] ?? this._returnFormatHandlers[ReturnFormat.VALUE_ONLY];
      const formattedChanges = format(changes);
      this._callback(formattedChanges);
```